### DISCUSSION TITLE
Protecting user's funds with covenants

### DISCUSSION THREAD

**Discussion by davidohayon669** [fedimint/fedimint#2599] [General]
This high level idea is based on my understanding of covenants, i'm sorry if it has holes in it due to not understanding the technicals entirely

If and when a version of covenants is deployed, specifically CTV which lets you pay to a merkle tree of unlimited size, a fedimint, or any other custodian model, can offer users to turn some of their funds into a trustlessly ejectable on-chain UTXO, reducing the trust in the custodian to only the "checking account"

Lets say someone starts to accept payments into a fedimint, after saving enough sats he can exchange a chunk of it into a UTXO in a covenant that the custodian creates periodically, that is locked to his private key for a specific duration. This can work exactly like a time-deposit in a bank, only with bitcoin a saver will be willing to PAY for the ability to have total control over this option to eject into base layer while also knowing his money is not rehypothecated. The setup is obviously trusting the custodian but a fedimint user trust their custodian anyway so it's only a benefit. A user can roll different funds in different durations and never expose more then a percent of his funds to a rug

This mechanism can lead to a balanced trade-off where 20% of your money is sitting with a custodian, available for fast and cheap payments, and 80% of it maintains full sovereignty, while the movement between your savings and checking account requires very small on-chain footprint and is managed by liquidity providers that earn a service fee

**Comment by justinmoon** [comment]
I'm very hazy on the details, but we've talked about doing something similar with [statechains](https://github.com/fedimint/fedimint/discussions/1502) in the past. It's an intriguing idea that deserves further exploration.

**Comment by humansinstitute** [comment]
ACK

I had similar thoughts a while back when working through the trade offs on [fedimint.org](https://fedimint.org/docs/TradeOffs/NotYourKeys) - Also the Ark proposal has some interesting parallels here as well. 

My initial concern with this was how to keep the UTXO tree up to date with the eCash notes as we'd need to update the UTXO commitments everytime the eCash notes were redeemed as I was assuming trying to do this one for one eCash to UTXO. 

However, I quite like this idea of having a portion of the funds locked in this manner which might simplify things. I'll give this one some thought as this would be a great area to have some more options.

**Comment by davidohayon669** [comment]
looks like some are already on to this
https://gist.github.com/moonsettler/6a214f5d01148ea204e9131b86a35382

**Reply by jkitman** [reply]
I've been following Ark lately, once that project gets fully reviewed and implemented I think it could be very promising as a Fedimint module https://github.com/fedimint/fedimint/discussions/2631

Alternatively we could use statechains to pass around large denominations of ecash.

**Comment by jkitman** [comment]
@davidohayon669 Trying to sum up similar ideas based on my limited understanding:

**Covenant-based pools** like [Tarpit](https://gist.github.com/moonsettler/6a214f5d01148ea204e9131b86a35382) could save on-chain fees, depending on how many other users participate when you enter/exit the covenant, but the drawback is you need other users to participate or you aren't really saving much and might as well just peg-out from the fedimint.

**Ark** #2631 allows for non-custodial batched transactions, but has the disadvantage of requiring a large number of users to be viable / users willing to participate in longer tx intervals.

**State-chains** #1502 allow users to transfer coins for free and redeem them non-custodially, but could still be rugged if the fed was always malicious (as opposed to coerced) and acted as one of the intermediary recipients.

All of these options give users more custodial power, but come with greater fees / less privacy as far as I can tell.  Another interesting point is that I think you can instantly swap into ecash or a payment channel with these constructions.
